### PR TITLE
Feat/transport server sni

### PR DESCRIPTION
**Pull request by j1m-ryan**
### Proposed changes

TransportServer SNI
#6324 

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

File: examples/custom-resources/transport-server-sni/README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+1. Create a GlobalConfiguration resource with the following listener.
+
+```yaml
+     listeners:
+      - name: tcp-listener
+        port: 7000 
+        protocol: TCP
+```
+
+1. Add a custom port to the NIC pod via the helm chart
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:14]:
```suggestion
2. Add a custom port to the NGINX Ingress Controller pod with the Helm chart.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+```yaml
+controller.customPorts:
+    - name: port
+      containerPort: 7000
+      protocol: TCP
+```
+
+1. Add a custom port to the NIC service
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:23]:
```suggestion
3. Add a custom port to the NGINX Ingress Controller service.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+```yaml
+controller.service.customPorts:
+          - name: tcp-port 
+            port: 7000 
+            protocol: TCP
+            targetPort: 7000 
+```
+
+1. Create the cafe-secret, and mongo-secret. These secrets are used for TLS in the TransportServers
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:33]:
```suggestion
4. Use `kubectl` to create the cafe-secret, and mongo-secret. These secrets are used for TLS in the TransportServers.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+`kubectl apply -f cafe-secret.yaml`
+`kubectl apply -f mongo-secret.yaml`
+
+1. Create the mongo and tcp echo example applications
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:37]:
```suggestion
5. Create the mongo and tcp echo example applications
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+`kubectl apply -f mongo.yaml`
+`kubectl apply -f tcp-echo-server.yaml`
+
+1. Wait until these are ready.
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:41]:
```suggestion
6. Wait until these are ready.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+`kubectl get deploy -w`
+
+1. Create the TransportServers for each application
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:44]:
```suggestion
7. Create the TransportServers for each application.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+`kubectl apply -f cafe-transport-server.yaml`
+`kubectl apply -f mongo-transport-server.yaml`
+
+1. Ensure they are in valid state
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:48]:
```suggestion
8. Ensure they are in valid state.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+`kubectl get ts`
+
+```bash
+NAME       STATE   REASON           AGE
+cafe-ts    Valid   AddedOrUpdated   2m
+mongo-ts   Valid   AddedOrUpdated   2m
+```
+
+1. Set up /etc/hosts or DNS
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:57]:
```suggestion
9. Set up /etc/hosts or DNS
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+In this example I am running NIC locally, so I will set
+cafe.example.com and mongo.example.com in my /etc/hosts to localhost
+If running NIC as a live service, the server names would be set at the DNS layer
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:60]:
```suggestion
This example uses a local NGINX Ingress Controller instance, so the /etc/hosts file
is being used to set cafe.example.com and mongo.example.com to localhost.
In a production instance, the server names would be set at the DNS layer.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+`cat /etc/hosts`
+
+```bash
+...
+127.0.0.1 cafe.example.com
+127.0.0.1 mongo.example.com
+```
+
+1. Expose port 7000 of the NIC Loadbalancer Service
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:69]:
```suggestion
10. Expose port 7000 of the LoadBalancer Service
```

In context, no need to keep mentioning the controller.

File: examples/custom-resources/transport-server-sni/README.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+`k port-forward svc/my-release-nginx-ingress-controller 7000:7000`
+
+1. Use ncat to ping cafe.example.com on port 7000 with ssl
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:72]:
```suggestion
11. Use ncat to ping cafe.example.com on port 7000 with ssl
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+`ncat --ssl cafe.example.com 7000`
+When you write a message you should get sent back
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:74]:
```suggestion
When you write a message you should receive the following response:
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+```bash
+hi
+hi
+```
+
+And once the connection is closed (ctrl + c) you should see the request and response in the NIC logs are both 2 bytes
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:81]:
```suggestion
Close the connection (CTRL+ c), then view the NGINX Ingress Controller logs.

The request and response should both be 2 bytes.
```

File: examples/custom-resources/transport-server-sni/README.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+```bash
+127.0.0.1 [24/Sep/2024:15:48:58 +0000] TCP 200 3 3 2.702 "-
+```
+
+1. Use mongosh to connect to the mongodb container, via the TransportServer, also on port 7000
```
Comment (ADubhlaoich) [inline examples/custom-resources/transport-server-sni/README.md:87]:
```suggestion
12. Use mongosh to connect to the mongodb container through the TransportServer on port 7000.
```

**Review by ADubhlaoich — APPROVED** [review]
LGTM, though I have a lot of edit suggestions.

Ordered list enumeration (1. repeated) is only available when writing a Hugo page.

It's very basic, but since explicit instructions are given for everything else, the example could also give the command to view the logs.

I may convert this to a proper documentation page at a later point: see #6572 for an example.

**Review by AlexFenlon — APPROVED** [review]

File: examples/custom-resources/transport-server-sni/README.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+## Add a custom port to the NGINX Ingress Controller service
+
+```yaml
+controller.service.customPorts:
+          - name: tcp-port 
+            port: 7000 
+            protocol: TCP
+            targetPort: 7000 
+```
```
Comment (pdabelf5) [inline examples/custom-resources/transport-server-sni/README.md:31]: consistent indentation of the yaml

File: examples/custom-resources/transport-server-sni/README.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+The request and response should both be 2 bytes.
+
+```bash
+127.0.0.1 [24/Sep/2024:15:48:58 +0000] TCP 200 3 3 2.702 "-
+```
+
+## Use mongosh to connect to the mongodb container through the TransportServer on port 7000
+
+`mongosh --host mongo.example.com --port 7000 --tls --tlsAllowInvalidCertificates`
```
Comment (pdabelf5) [inline examples/custom-resources/transport-server-sni/README.md:99]: should we advise users that they need to install `mongosh` ?

File: examples/custom-resources/transport-server-sni/README.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+127.0.0.1 mongo.example.com
+```
+
+## Expose port 7000 of the LoadBalancer service
+
+`kubectl port-forward svc/my-release-nginx-ingress-controller 7000:7000`
+
+## Use `ncat` to ping cafe.example.com on port 7000 with SSL
+
+`ncat --ssl cafe.example.com 7000`
```
Comment (pdabelf5) [inline examples/custom-resources/transport-server-sni/README.md:81]: Netcat binary is `nc` by default?

File: examples/custom-resources/transport-server-sni/tcp-echo-server.yaml [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+       containers:
+       - name: tcp-echo-server
+         image: alpine
+         command: ["/bin/sh"]
+         args:
+         - -c
+         - |
+           apk add --no-cache socat
+           echo "TCP Echo Server starting..."
+           socat -v tcp-listen:7000,fork,reuseaddr,keepalive exec:'/bin/cat'
```
Comment (pdabelf5) [inline examples/custom-resources/transport-server-sni/tcp-echo-server.yaml:24]: do we need `socat` here?  Alpine comes with netcat (`nc`) already

File: internal/configs/version2/__snapshots__/templates_test.snap [hunk 19]
```diff
@@ -101,6 +161,7 @@ server {
     listen 127.0.0.1:1234 ssl udp;
     listen [::1]:1234 ssl udp;
 
+
```
Comment (pdabelf5) [inline internal/configs/version2/__snapshots__/templates_test.snap:164]: Are these newlines needed?

File: examples/custom-resources/transport-server-sni/tcp-echo-server.yaml (hunk 18 above)
Comment (j1m-ryan) [reply examples/custom-resources/transport-server-sni/tcp-echo-server.yaml]: No. Thanks this is better will update.

File: examples/custom-resources/transport-server-sni/README.md (hunk 16 above)
Comment (j1m-ryan) [reply examples/custom-resources/transport-server-sni/README.md:102]: Yeah that makes sense. I've added instructions to the top of the reademe now for `ncat` (from nmap) and `mongosh` from the mongo docs.

File: examples/custom-resources/transport-server-sni/README.md (hunk 15 above)
Comment (j1m-ryan) [reply examples/custom-resources/transport-server-sni/README.md]: Missed this thanks

File: internal/configs/version2/__snapshots__/templates_test.snap (hunk 19 above)
Comment (j1m-ryan) [reply internal/configs/version2/__snapshots__/templates_test.snap]: They're not no. Updated in latest commit

**Review by jjngx — APPROVED** [review]
🚀

**Review by pdabelf5 — APPROVED** [review]

**Pull request merged**